internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -269,67 +241,161 @@ func writeKubeadmConfig(data *userdata.UserData) (err error) {
+	certs := []struct {
+		Cert     *x509.PEMEncodedCertificateAndKey
+		CertPath string
+		KeyPath  string
+	}{
+		{
+			Cert:     data.Security.Kubernetes.CA,
+			CertPath: constants.KubeadmCACert,
+			KeyPath:  constants.KubeadmCAKey,
+		},
+		{
+			Cert:     data.Security.Kubernetes.SA,
+			CertPath: constants.KubeadmSACert,
+			KeyPath:  constants.KubeadmSAKey,
+		},
+		{
+			Cert:     data.Security.Kubernetes.FrontProxy,
+			CertPath: constants.KubeadmFrontProxyCACert,
+			KeyPath:  constants.KubeadmFrontProxyCAKey,
+		},
+		{
+			Cert:     data.Security.Kubernetes.Etcd,
+			CertPath: constants.KubeadmEtcdCACert,
+			KeyPath:  constants.KubeadmEtcdCAKey,
+		},
 	}
-	if data.Crt != nil {
 
-		if err = os.MkdirAll(path.Dir(cert), 0600); err != nil {
-			return err
+	for _, cert := range certs {
+		if cert.Cert == nil {
+			continue
+		}
+
+		if cert.Cert.Crt != nil {
+			if err = os.MkdirAll(path.Dir(cert.CertPath), 0600); err != nil {
+				return err
+			}
+			if err = ioutil.WriteFile(cert.CertPath, cert.Cert.Crt, 0400); err != nil {
+				return fmt.Errorf("write %s: %v", cert.CertPath, err)
+			}
 		}
-		if err = ioutil.WriteFile(cert, data.Crt, 0400); err != nil {
-			return fmt.Errorf("write %s: %v", cert, err)
+		if cert.Cert.Key != nil {
+			if err = os.MkdirAll(path.Dir(cert.KeyPath), 0600); err != nil {
+				return err
+			}
+			if err = ioutil.WriteFile(cert.KeyPath, cert.Cert.Key, 0400); err != nil {
+				return fmt.Errorf("write %s: %v", cert.KeyPath, err)
+			}
 		}
 	}
-	if data.Key != nil {
-		if err = os.MkdirAll(path.Dir(key), 0600); err != nil {
-			return err
-		}
-		if err = ioutil.WriteFile(key, data.Key, 0400); err != nil {
-			return fmt.Errorf("write %s: %v", key, err)
+	return err
+}
+
+// files sets up the routines to pull the missing
+// required files from its peers
+func files() []*proto.ReadFileRequest {
+	files := []string{
+		constants.AuditPolicyPathInitramfs,
+		constants.EncryptionConfigInitramfsPath,
+		constants.KubeadmCACert,
+		constants.KubeadmCAKey,
+		constants.KubeadmSACert,
+		constants.KubeadmSAKey,
+		constants.KubeadmFrontProxyCACert,
+		constants.KubeadmFrontProxyCAKey,
+		constants.KubeadmEtcdCACert,
+		constants.KubeadmEtcdCAKey,
+	}
+
+	fileRequests := []*proto.ReadFileRequest{}
+	// Check to see if we already have the file locally
+	for _, file := range files {
+		if _, err := os.Stat(file); os.IsNotExist(err) {
+			fileRequests = append(fileRequests, &proto.ReadFileRequest{Path: file})
 		}
 	}
-	return nil
+
+	return fileRequests
 }
 
-func writeFiles(client proto.TrustdClient, files []string) (err error) {
-	errChan := make(chan error)
-	doneChan := make(chan bool)
-	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Minute)
-	defer cancelFunc()
+// createTrustdClients handles instantiating a trustd client connection
+// to each trustd endpoint defined in userdata
+func createTrustdClients(data *userdata.UserData) ([]proto.TrustdClient, error) {
```

Comment:
Perhaps a named return here would be better:

```
func createTrustdClients(data *userdata.UserData) (trustds []proto.TrustdClient, err error)
```

Reply:
thought there was concern over naked returns?

Reply:
A named return doesn't require a naked return. Naked returns are generally frowned upon, but named returns are not.